[LOOP-2140] refactored workout override reminder

nhamming opened the pull request:
https://tidepool.atlassian.net/browse/LOOP-2140

workout override reminder now triggers with loop cycle and will not fire when app is killed

Just FYI, I looked at alternatives to including alerting in the `LoopDataManager`, which included separating the logic out of `LoopSettings`, but this was complicated by the `LoopCore` framework and its benefit didn't seem worth the effort. Thus I landed on this simpler solution, but it does muddy the `LoopDataManager` waters.

nhamming committed:
workout override reminder now triggers with loop cycle and will not fire when app is killed

nhamming committed:
clean up

nhamming committed:
more clean up

nhamming committed:
last of the clean up (hopefully)

nhamming committed:
transfer the enabled date when set on the watch

rickpasetto commented on Loop/Managers/LoopDataManager.swift:
```diff
@@ -707,6 +710,7 @@ extension LoopDataManager {
             self.logger.default("Loop running")
             NotificationCenter.default.post(name: .LoopRunning, object: self)
 
+            self.checkAlerts()
```
Why not have something that listens to `.LoopRunning` (above)?

nhamming replied:
see my other response

rickpasetto reviewed (commented):
Hm.  Seems to me you could create a separate `class` that only takes an `AlertPresenter` and `LoopSettings` that listens to `.LoopRunning`.  Still not seeing why this needs to be added to `LoopDataManager`.  What am I missing?

nhamming commented:
> Hm. Seems to me you could create a separate `class` that only takes an `AlertPresenter` and `LoopSettings` that listens to `.LoopRunning`. Still not seeing why this needs to be added to `LoopDataManager`. What am I missing?

@rickpasetto This is what I was hinting at in the description. Encapsulating `LoopSettings` inside a `LoopSettingsManager` requires `LoopSettingsManager` to be put in the `LoopCore` to make it accessible to the watch. The watch doesn't really need a `LoopSettingsManager`. However, the logic could be abstracted from the `LoopSettings` model (which I think should be done). Going down the path of abstracting the logic from the model lead to other complications and that is when I pulled the plug on this approach since it was creating a large number of changes and potential complexities in the code when the simpler solution above solves the problem.

Also, I wonder about the need of the `LoopDataManager` to trigger other alerts in addition to just `LoopSettings`.

Let me know what you think. This code will be left open for a few days as I'm back on 510k tickets. But once I'm freed up, I'll come back to this and check in with you (if there is no response).

rickpasetto commented:
@nhamming I still don't follow this.  The class I'm proposing would simply be a peer of `LoopDataManager`, owned by it.  I don't propose "encapsulating `LoopSettings` into a `LoopSettingsManager`" at all.  Something like this (not actual code):

```swift
protocol LoopAlertCheckerDelegate {
    var settings: LoopSettings { get }
}
class LoopAlertChecker {
    weak var delegate: LoopAlertCheckerDelegate?
    var alertPresenter: AlertPresenter
    init(delegate: LoopAlertCheckerDelegate, alertPresenter: AlertPresenter) {
        self.delegate = delegate
        self.alertPresenter = alertPresenter
    }
    
    private func checkAlerts() {
        checkWorkoutOverrideReminder()
    }

    private func checkWorkoutOverrideReminder() {
        guard delegate?.settings.isScheduleOverrideInfiniteWorkout else {
            delegate?.settings.indefiniteWorkoutOverrideEnabledDate = nil
            return
        }
...
}

    private func issueWorkoutOverrideReminder() {
        alertPresenter.issueAlert(settings.workoutOverrideReminderAlert)
    }
}
```

WDYT?

nhamming committed:
Merge branch 'dev' into nate/task/LOOP-2140/workout-override-reminder-refactor

nhamming commented:
> @nhamming I still don't follow this. The class I'm proposing would simply be a peer of `LoopDataManager`, owned by it. I don't propose "encapsulating `LoopSettings` into a `LoopSettingsManager`" at all. Something like this (not actual code):
> 
> ```swift
> protocol LoopAlertCheckerDelegate {
>     var settings: LoopSettings { get }
> }
> class LoopAlertChecker {
>     weak var delegate: LoopAlertCheckerDelegate?
>     var alertPresenter: AlertPresenter
>     init(delegate: LoopAlertCheckerDelegate, alertPresenter: AlertPresenter) {
>         self.delegate = delegate
>         self.alertPresenter = alertPresenter
>     }
>     
>     private func checkAlerts() {
>         checkWorkoutOverrideReminder()
>     }
> 
>     private func checkWorkoutOverrideReminder() {
>         guard delegate?.settings.isScheduleOverrideInfiniteWorkout else {
>             delegate?.settings.indefiniteWorkoutOverrideEnabledDate = nil
>             return
>         }
> ...
> }
> 
>     private func issueWorkoutOverrideReminder() {
>         alertPresenter.issueAlert(settings.workoutOverrideReminderAlert)
>     }
> }
> ```
> 
> WDYT?

Ah... I see. I was coming at this from a different approach based on our past discussion about LoopSettings containing policy that maybe it shouldn't. But I'll take this alternative.

nhamming commented:
@rickpasetto one problem it this delegate approach is that the settings parameters need to be updated. I'll push the encapsulated approach to get your thoughts on it.

nhamming committed:
trying a different approach

nhamming committed:
clean up

nhamming committed:
more clean up

rickpasetto commented on Loop/Managers/LoopSettingsManager.swift:
```diff
@@ -0,0 +1,56 @@
+    //TODO add tests
+    private let alertManager: AlertPresenter?
+
+    @Published var settings: LoopSettings
```
Hm...not...quite what I was thinking, encapsulating `LoopSettings` in this way.  But I guess I see your approach.

nhamming replied:
After a night to sleep on it, I'm going to take your suggestion. I realized I was overlooking using the protocol to also set. 

thanks for your thoughts.

rickpasetto commented on LoopTests/Managers/LoopDataManagerAlertingTests.swift:
```diff
@@ -0,0 +1,103 @@
+import LoopCore
+@testable import Loop
+
+class LoopDataManagerAlertingTests: XCTestCase {
```
Now you can refactor these tests to test `LoopSettingsManager`, right?

rickpasetto commented on LoopCore/LoopSettings.swift:
```diff
@@ -33,6 +24,10 @@ public struct LoopSettings {
 
+    public var indefiniteWorkoutOverrideEnabledDate: Date?
+
+    public var workoutOverrideReminderInterval: TimeInterval = .days(1)
```
Consider changing this to a `let`

rickpasetto commented on LoopCore/LoopSettings.swift:
```diff
@@ -181,16 +174,15 @@ extension LoopSettings {
     }
 
-    public func legacyWorkoutOverride(beginningAt date: Date = Date(), for duration: TimeInterval) -> TemporaryScheduleOverride? {
+    public mutating func legacyWorkoutOverride(beginningAt date: Date = Date(), for duration: TimeInterval) -> TemporaryScheduleOverride? {
```
Instead of having to make this `mutating`, can you move the `indefiniteWorkoutOverrideEnabledDate` state into your new `LoopSettingsManager`?

nhamming replied:
After sleep, I decided to follow your original proposal. Given that, the LoopSettingsAlerter won't track the usage of the settings nor do I think it should handle storing/restoring. What do you think?

rickpasetto commented on LoopCore/LoopSettings.swift:
```diff
@@ -181,16 +174,15 @@ extension LoopSettings {
     }
 
-    public func legacyWorkoutOverride(beginningAt date: Date = Date(), for duration: TimeInterval) -> TemporaryScheduleOverride? {
+    public mutating func legacyWorkoutOverride(beginningAt date: Date = Date(), for duration: TimeInterval) -> TemporaryScheduleOverride? {
```
Instead of having to make this `mutating`, can you move the `indefiniteWorkoutOverrideEnabledDate` state into your new `LoopSettingsManager`?

rickpasetto replied:
Seems a bit odd to me to still have `LoopSettings` have this responsibility yet have `LoopSettingsManager` issue the reminder.  Do you agree?

rickpasetto commented on Loop/Managers/LoopSettingsManager.swift:
```diff
@@ -0,0 +1,56 @@
+        self.alertManager = alertManager
+        self.settings = settings
+
+        NotificationCenter.default.addObserver(forName: .LoopRunning,
```
[nit] consider removing the extraneous vertical and whitespace here, i.e.
```
        NotificationCenter.default.addObserver(forName: .LoopRunning, object: nil, queue: nil) {
```

rickpasetto reviewed (commented):
Added some comments...LMK what you think...

nhamming committed:
reducing to a loop settings alerter

nhamming committed:
putting files in more discoverable locations

nhamming committed:
responses to PR comments

nhamming committed:
workout override enabled date is already reset in LoopSettings

nhamming commented:
@rickpasetto I believe this addresses all your comments except the who is responsible for handling the enabled date of the workout override (see my reply).

rickpasetto reviewed (approved):
Nice...LGTM.  Thanks for working with me on this! 👍

nhamming committed:
clean up

nhamming committed:
Merge branch 'dev' into nate/task/LOOP-2140/workout-override-reminder-refactor
